gitbook-docs/request_response.md

```diff
@@ -0,0 +1,130 @@
+# Request/Response
+
+## WebSockets and similar transports
+
+The exact format of the message for a specific request is defined elsewhere in the specification.
+
+A request should include the `context` when appropriate and must include a client generated `requestId`. The `requestId` is a string and it must be a [v4 UUID](https://tools.ietf.org/html/rfc4122.html#section-4.4). It will always be included in any response to the request by the server.
+
+
+For example. a request to PUT a value:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "put": {
+    "path": "electrical.switches.anchorLight.state",
+    "value": 1
+  }
+}
+```
+
+The server will respond with a message which includes the `requestId`, `correlationId` if provided, and a `state`.
+
+The `state` can be `PENDING` or `COMPLETED`
+
+When the state is `COMPLETED`, the message will contain a `result` value. The `result` will be any standard HTTP code including the following.
+
+- 200 - the request was succesfull
+- 502 - something went wrong carrying out the request on the server side
+- 400 - something is wrong with the client's request
+- 504 - timeout on the server side trying to carry out the request
+- 405 - the server does not support the request
+- 401 - the request has not been applied because it lacks valid authentication credentials
+- 403 - the client does not have permission to make the request
+
+The message can optionally contain a `message`.
+
+The response object may contain other objects depending on the specific request being made. For example, a response to auth request could contain a `login` object.
```

Comment:
Wouldn't it be easier for implementors to have a single, optional key `payload` which can be used to convey information such as this example? That way the client doesn't need to use a different key every single time, which is cumbersome in my experience.

Reply:
I initially had `response` instead of payload, but @rob42 requested that we have keys for specific requests.